Feature/events

Hackatosh opened the pull request:
- Added custom validators in **utils.ts**
- Added _PetEvent_ model to store events in DB in **event.ts** (and modified **initDB.ts** and **associations.ts**)
- Deleted unused timestamp in models and fixed associations
- Created an events router in **events.ts** which allows to :
-> Create a new event
-> Get an event by eventId
-> Modifiy an event by eventId
->Delete an event by eventId
-> Add an animal to an event
-> Delete an animal from an event
Event can be modified only by the user who created them and can be then accessed by everyone logged in. An animal can be added and deleted only by its owner.

**To test :** 
- Register a user and log in
- Add an animal for this user
- Create an event (with _POST_ on _/events/_) and then access its informations (with _GET_ on _/events/:eventId_)
- Modify the event (with _PUT_ on _/events/:eventId_) and see the modifications (with _GET_ on _/events/:eventId_)
- Add an animal (with _POST_ on _/events/:eventId/animals_) and see the animal added (with _GET_ on _/events/:eventId_)
- Delete the animal (with _DELETE_ on _/events/:eventId/animals/:animalId_) and see that the animal has been deleted (with _GET_ on _/events/:eventId_)
- Delete the event (with _DELETE_ on _/events/:eventId_) and see that it has been deleted (with _GET_ on _/events/:eventId_)

apollisa commented:
Found two bugs but otherwise this is really cool!

### Bugs
- When creating an event, `endDate` can be before `startDate`, but should not.
- When creating an event with an associated species, the EventSpecies table is not populated. (The same goes when updating an event.)

### Suggestions
- Perhaps none of the fields of `Event` should be allowed to be null? At least not `userId` and `location`?
- Perhaps dates should really be datetimes, because an event could run from an hour to another, not from a day to another.
- I think ‘specie’ is incorrect and should always be ‘species.’

Hackatosh commented:
**Bugs fixed:**
- Added validation to compare beginDate and endDate
- EventSpcies seems to be populated. Could you be more precise ?

- userId is allowed to be null to allow importation from outside the application. Location is null to allow the creation of event not related to a place (e.g. : "National Dog Day")
- Converted Date to DateTime when needed
- Although "Specie" is mispelled, it allows to distinguish the Model (Specie) from the tablename (species). This should be kept to avoid confusion.

Hackatosh commented:
**Bugs fixed :**
- Added validation to compare beginDate and endDate
- EventSpcies seems to be populated. Could you be more precise ?

**Suggestions :**
- userId is allowed to be null to allow importation from outside the application. Location is null to allow the creation of event not related to a place (e.g. : "National Dog Day")
- Converted Date to DateTime when needed
- Although "Specie" is mispelled, it allows to distinguish the Model (Specie) from the tablename (species). This should be kept to avoid confusion.

apollisa commented on back/src/core/utils.ts:
```diff
@@ -24,8 +26,46 @@ const objectifyHeadersArray = function (headers:Array<HttpHeader>):OutgoingHttpH
+
+/*** Check if a string is a valid datetime***/
+const isDateTimeValid = function (date:string): boolean {
+    return moment(date, 'DD/MM/YYYY:hh:mm:ss',true).isValid();
```
Why not use `moment(date).isValid()`, as according to the docs it will correctly parse an ISO 8601, which is what an `<input type="date">` and `<input type="time">` will return? At the very least you should change the ‘h’ into ‘H’, otherwise the hours are expected to be between 1 and 24, and not 0 and 23.

Hackatosh replied:
This has been fixed, thanks for the input !
